GeographicLib-2.1/man/MagneticField.1
.\" Automatically generated by Pod::Man 4.14 (Pod::Simple 3.43)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{\
.    if \nF \{\
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{\
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "MAGNETICFIELD 1"
.TH MAGNETICFIELD 1 "2022-05-06" "GeographicLib 2.1" "GeographicLib Utilities"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
MagneticField \-\- compute the earth's magnetic field
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBMagneticField\fR [ \fB\-n\fR \fIname\fR ] [ \fB\-d\fR \fIdir\fR ]
[ \fB\-N\fR \fINmax\fR ] [ \fB\-M\fR \fIMmax\fR ]
[ \fB\-t\fR \fItime\fR | \fB\-c\fR \fItime\fR \fIlat\fR \fIh\fR ]
[ \fB\-r\fR ] [ \fB\-w\fR ] [ \fB\-T\fR \fItguard\fR ] [ \fB\-H\fR \fIhguard\fR ] [ \fB\-p\fR \fIprec\fR ]
[ \fB\-v\fR ]
[ \fB\-\-comment\-delimiter\fR \fIcommentdelim\fR ]
[ \fB\-\-version\fR | \fB\-h\fR | \fB\-\-help\fR ]
[ \fB\-\-input\-file\fR \fIinfile\fR | \fB\-\-input\-string\fR \fIinstring\fR ]
[ \fB\-\-line\-separator\fR \fIlinesep\fR ]
[ \fB\-\-output\-file\fR \fIoutfile\fR ]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBMagneticField\fR reads in times and positions on standard input and
prints out the geomagnetic field on standard output and, optionally, its
rate of change.
.PP
The input line is of the form \fItime\fR \fIlat\fR \fIlon\fR \fIh\fR. \fItime\fR is a
date of the form 2012\-11\-25 (yyyy-mm-dd or yyyy-mm), a fractional year
such as 2012.9, or the string \*(L"now\*(R".  \fIlat\fR and \fIlon\fR are the
latitude and longitude expressed as decimal degrees or degrees,
minutes, and seconds; for details on the allowed formats for latitude
and longitude, see the \f(CW\*(C`GEOGRAPHIC COORDINATES\*(C'\fR section of
\&\fBGeoConvert\fR\|(1).  \fIh\fR is the height above the ellipsoid in meters; this
is optional and defaults to zero.  Alternatively, \fItime\fR can be given
on the command line as the argument to the \fB\-t\fR option, in which case
it should not be included on the input lines.  Finally, the magnetic
field can be computed at various points on a circle of latitude
(constant \fItime\fR, \fIlat\fR, and \fIh\fR) via the \fB\-c\fR option; in this
case only the longitude should be given on the input lines.
.PP
The output consists of the following 7 items:
.PP
.Vb 9
\&  the declination (the direction of the horizontal component of
\&    the magnetic field measured clockwise from north) in degrees,
\&  the inclination (the direction of the magnetic field measured
\&    down from the horizontal) in degrees,
\&  the horizontal field in nanotesla (nT),
\&  the north component of the field in nT,
\&  the east component of the field in nT,
\&  the vertical component of the field in nT (down is positive),
\&  the total field in nT.
.Ve
.PP
If the \fB\-r\fR option is given, a second line is printed giving the rates
of change of these quantities in degrees/yr and nT/yr.
.PP
The \s-1WGS84\s0 ellipsoid is used, \fIa\fR = 6378137 m, \fIf\fR = 1/298.257223563.
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\fB\-n\fR \fIname\fR" 4
.IX Item "-n name"
use magnetic field model \fIname\fR instead of the default \f(CW\*(C`wmm2020\*(C'\fR.  See
\&\*(L"\s-1MODELS\*(R"\s0.
.IP "\fB\-d\fR \fIdir\fR" 4
.IX Item "-d dir"
read magnetic models from \fIdir\fR instead of the default.  See
\&\*(L"\s-1MODELS\*(R"\s0.
.IP "\fB\-N\fR \fINmax\fR" 4
.IX Item "-N Nmax"
limit the degree of the model to \fINmax\fR.
.IP "\fB\-M\fR \fIMmax\fR" 4
.IX Item "-M Mmax"
limit the order of the model to \fIMmax\fR.
.IP "\fB\-t\fR \fItime\fR" 4
.IX Item "-t time"
evaluate the field at \fItime\fR instead of reading the time from the input
lines.
.IP "\fB\-c\fR \fItime\fR \fIlat\fR \fIh\fR" 4
.IX Item "-c time lat h"
evaluate the field on a circle of latitude given by \fItime\fR, \fIlat\fR,
\&\fIh\fR instead of reading these quantities from the input lines.  In this
case, \fBMagneticField\fR can calculate the field considerably more
quickly.
.IP "\fB\-r\fR" 4
.IX Item "-r"
toggle whether to report the rates of change of the field.
.IP "\fB\-w\fR" 4
.IX Item "-w"
toggle the longitude first flag (it starts off); if the flag is on, then
on input and output, longitude precedes latitude (except that, on input,
this can be overridden by a hemisphere designator, \fIN\fR, \fIS\fR, \fIE\fR,
\&\fIW\fR).
.IP "\fB\-T\fR \fItguard\fR" 4
.IX Item "-T tguard"
signal an error if \fItime\fR lies \fItguard\fR years (default 50 yr) beyond
the range for the model.
.IP "\fB\-H\fR \fIhguard\fR" 4
.IX Item "-H hguard"
signal an error if \fIh\fR lies \fIhguard\fR meters (default 500000 m) beyond
the range for the model.
.IP "\fB\-p\fR \fIprec\fR" 4
.IX Item "-p prec"
set the output precision to \fIprec\fR (default 1).  Fields are printed
with precision with \fIprec\fR decimal places; angles use \fIprec\fR + 1
places.
.IP "\fB\-v\fR" 4
.IX Item "-v"
print information about the magnetic model on standard error before
processing the input.
.IP "\fB\-\-comment\-delimiter\fR \fIcommentdelim\fR" 4
.IX Item "--comment-delimiter commentdelim"
set the comment delimiter to \fIcommentdelim\fR (e.g., \*(L"#\*(R" or \*(L"//\*(R").  If
set, the input lines will be scanned for this delimiter and, if found,
the delimiter and the rest of the line will be removed prior to
processing and subsequently appended to the output line (separated by a
space).
.IP "\fB\-\-version\fR" 4
.IX Item "--version"
print version and exit.
.IP "\fB\-h\fR" 4
.IX Item "-h"
print usage, the default magnetic path and name, and exit.
.IP "\fB\-\-help\fR" 4
.IX Item "--help"
print full documentation and exit.
.IP "\fB\-\-input\-file\fR \fIinfile\fR" 4
.IX Item "--input-file infile"
read input from the file \fIinfile\fR instead of from standard input; a file
name of \*(L"\-\*(R" stands for standard input.
.IP "\fB\-\-input\-string\fR \fIinstring\fR" 4
.IX Item "--input-string instring"
read input from the string \fIinstring\fR instead of from standard input.
All occurrences of the line separator character (default is a semicolon)
in \fIinstring\fR are converted to newlines before the reading begins.
.IP "\fB\-\-line\-separator\fR \fIlinesep\fR" 4
.IX Item "--line-separator linesep"
set the line separator character to \fIlinesep\fR.  By default this is a
semicolon.
.IP "\fB\-\-output\-file\fR \fIoutfile\fR" 4
.IX Item "--output-file outfile"
write output to the file \fIoutfile\fR instead of to standard output; a
file name of \*(L"\-\*(R" stands for standard output.
.SH "MODELS"
.IX Header "MODELS"
\&\fBMagneticField\fR computes the geomagnetic field using one of the
following models
.PP
.Vb 10
\&    wmm2010, the World Magnetic Model 2010, which approximates the
\&      main magnetic field for the period 2010\-2015.  See
\&      https://ngdc.noaa.gov/geomag/WMM/DoDWMM.shtml
\&    wmm2015v2, the World Magnetic Model 2015, which approximates the
\&      main magnetic field for the period 2015\-2020.  See
\&      https://ngdc.noaa.gov/geomag/WMM/DoDWMM.shtml
\&    wmm2015, a deprecated version of wmm2015v2
\&    wmm2020, the World Magnetic Model 2020, which approximates the
\&      main magnetic field for the period 2020\-2025.  See
\&      https://ngdc.noaa.gov/geomag/WMM/DoDWMM.shtml
\&    igrf11, the International Geomagnetic Reference Field (11th
\&      generation), which approximates the main magnetic field for
\&      the period 1900\-2015.  See
\&      https://ngdc.noaa.gov/IAGA/vmod/igrf.html
\&    igrf12, the International Geomagnetic Reference Field (12th

ros_env/history.txt
 1071  rostopic list
 1072  sshm
 1073  roslaunch omo_r1mini_bringup omo_r1_imu.launch
 1074  cm
 1075  roslaunch omo_r1mini_bringup omo_r1_imu.launch
 1076  roslaunch omo_r1_teleop omo_r1_teleop_key.launch 
 1077  roscore
 1078  nb
 1079  eb
 1080  roscore
 1081  sshr
 1082  sshr
 1083  anydesk
 1084  roscore
 1085  sshm
 1086  sudo apt install ros-melodic-cartographer-ros
 1087  roscore
 1088  gazebo
 1089  rviz
 1090  sshm
 1091  roscore
 1092  roslaunch omo_r1mini_cartographer omo_r1mini_cartographer_rviz.launch
 1093  roslaunch omo_r1mini_navigation omo_r1mini_navigation_rviz.launc
 1094  rosclean
 1095  roslaunch omo_r1mini_navigation omo_r1mini_navigation_rviz.launc
 1096  roslaunch omo_r1mini_navigation omo_r1mini_navigation_rviz.launch
 1097  roslaunch omo_r1mini_teleop omo_r1mini_teleop_key.launch
 1098  cd {$HOME}
 1099  hp,e
 1100  cd home
 1101  sshm
 1102  sshm
 1103  cd to map folder
 1104  cs
 1105  cd ..
 1106  ls
 1107  cd map
 1108  rosrun map_server map_saver -f adyslam
 1109  roscore
 1110  roslaunch omo_r1mini_cartographer omo_r1mini_cartographer_rviz.launch
 1111  roslaunch omo_r1mini_teleop omo_r1mini_teleop_key.launch
 1112  roscore
 1113  roslaunch omo_r1mini_navigation omo_r1mini_navigation_rviz.launch
 1114  sshm
 1115  sshm
 1116  roscore
 1117  roslaunch omo_r1mini_navigation omo_r1mini_navigation_rviz.launch
 1118  roslaunch omo_r1mini_navigation omo_r1mini_navigation.launch
 1119  rviz
 1120  roscore
 1121  rviz
 1122  roscore
 1123  sshm
 1124  roslaunch omo_r1mini_navigation omo_r1mini_navigation_rviz.launch
 1125  sshm
 1126  roscore
 1127  roslaunch omo_r1mini_navigation omo_r1mini_navigation_rviz.launch
 1128  sshm1
 1129  sshm
 1130  roscore
 1131  roslaunch omo_r1mini_navigation omo_r1mini_navigation_rviz.launch
 1132  sudo systemctl restart t
 1133  sshm
 1134  sshm
 1135  roscore
 1136  roslaunch omo_r1mini_navigation omo_r1mini_navigation_rviz.launch
 1137  roscore
 1138  sshm
 1139  rosrun map_server map_server map_file.yaml
 1140  [출처] ROS Navigation Map Creation|작성자 Tomboy
 1141  rosrun map_server map_server map.yaml
 1142  rosrun map_server map_server adyslam.yaml
 1143  roscore
 1144  cs
 1145  cd omo_r1mini/omo_r1mini_navigation/maps
 1146  rosrun map_server map_saver -f adyslam2
 1147  ./catkin_ws/src/omo_r1mini/omo_r1mini_cartographer/scripts/save_pbstream.sh
 1148  roslaunch omo_r1mini_navigation omo_r1mini_navigation_rviz.launch
 1149  roslaunch omo_r1mini_cartographer omo_r1mini_cartographer_rviz.launch
 1150  roslaunch omo_r1mini_navigation omo_r1mini_navigation_rviz.launch
 1151  roslaunch omo_r1mini_teleop omo_r1mini_teleop_key.launch
 1152  sshm
 1153  roscore
 1154  roslaunch omo_r1mini_navigation omo_r1mini_navigation_rviz.launch
 1155  roscore
 1156  rosrun map_server map_saver -f map.yaml
 1157  rosrun map_server map_saver -f map
 1158  roscore
 1159  sshm
 1160  roscore
 1161  roslaunch omo_r1mini_cartographer omo_r1mini_cartographer_rviz.launch
 1162  roslaunch omo_r1mini_teleop omo_r1mini_teleop_key.launc
 1163  roslaunch omo_r1mini_teleop omo_r1mini_teleop_key.launch
 1164  roslaunch omo_r1mini_navigation omo_r1mini_navigation_rviz.launch
 1165  cs
 1166  cd omo_r1mini/omo_r1mini_navigation/maps
 1167  rosrun map_server map_saver -f tutorial
 1168  roscore
 1169  roslaunch omo_r1mini_navigation omo_r1mini_navigation_rviz.launch
 1170  rqt_graph
 1171  sshm
 1172  sshm
 1173  sshm
 1174  roscore
 1175  sshm
 1176  roslaunch omo_r1mini_slam omo_r1mini_slam_rviz.launch
 1177  roslaunch omo_r1mini_teleop omo_r1mini_teleop_key.launch
 1178  sshm
 1179  sshm
 1180  sshm
 1181  sshm
 1182  roscore
 1183  roslaunch omo_r1mini_navigation omo_r1mini_navigation_rviz.launch
 1184  roslaunch omo_r1mini_teleop omo_r1mini_teleop_key.launch
 1185  roslaunch omo_r1mini_slam omo_r1mini_slam_rviz.launch
 1186  roscore
 1187  sshm
 1188  sshm
 1189  roscore
 1190  roslaunch omo_r1mini_slam omo_r1mini_slam_rviz.launch
 1191  roslaunch omo_r1mini_teleop omo_r1mini_teleop_key.launch
 1192  roscore
 1193  rosrun map_server map_server map.yaml
 1194  roslaunch omo_r1mini_slam omo_r1mini_slam_rviz.launch
 1195  roscore
 1196  roslaunch omo_r1mini_slam omo_r1mini_slam_rviz.launch
 1197  rosrun map_server map_server map.yaml
 1198  roslaunch omo_r1mini_slam omo_r1mini_slam_rviz.launch
 1199  cp map.*home/
 1200  sshm
 1201  sshm
 1202  roscore
 1203  roslaunch omo_r1mini_slam omo_r1mini_slam_rviz.launch
 1204  roslaunch omo_r1mini_navigation omo_r1mini_navigation_rviz.launch
 1205  sshm
 1206  sshm
 1207  roscore
 1208  roslaunch omo_r1mini_navigation omo_r1mini_navigation_rviz.launch
 1209  roslaunch omo_r1mini_navigation omo_r1mini_navigation_rviz.launch
 1210  sshm
 1211  roscore
 1212  sshm
 1213  roslaunch omo_r1mini_navigation omo_r1mini_navigation_rviz.launch
 1214  roslaunch omo_r1mini_slam omo_r1mini_slam_rviz.launch
 1215  sshm
 1216  sshm
 1217  ls
 1218  sshm
 1219  roscore
 1220  sshm 
 1221  sshm
 1222  gogo
 1223  sshm
 1224  roscore
 1225  roslaunch omo_r1mini_navigation omo_r1mini_navigation_rviz.launch
 1226  sshm
 1227  sshm
 1228  roscore
 1229  roslaunch omo_r1mini_navigation omo_r1mini_navigation_rviz.launch
 1230  sshm
 1231  roscore
 1232  roslaunch omo_r1mini_navigation omo_r1mini_navigation_rviz.launch
 1233  sshm
 1234  roscore
 1235  sshm1
 1236  sshm
 1237  roscore
 1238  roslaunch omo_r1mini_navigation omo_r1mini_navigation_rviz.launch
 1239  sshm
 1240  sshm
 1241  git clone nav2_bt_navigator/behavior_trees/navigate_through_poses_w_replanning_and_recovery.xml

GeographicLib-1.34/tools/MagneticField.cpp
/**
 * \file MagneticField.cpp
 * \brief Command line utility for evaluating magnetic fields
 *
 * Copyright (c) Charles Karney (2011-2012) <charles@karney.com> and licensed
 * under the MIT/X11 License.  For more information, see
 * http://geographiclib.sourceforge.net/
 *
 * Compile and link with
 *   g++ -g -O3 -I../include -I../man -o MagneticField \
 *       MagneticField.cpp \
 *       ../src/CircularEngine.cpp \
 *       ../src/DMS.cpp \
 *       ../src/Geocentric.cpp \
 *       ../src/MagneticCircle.cpp \
 *       ../src/MagneticModel.cpp \
 *       ../src/SphericalEngine.cpp \
 *       ../src/Utility.cpp
 *
 * See the <a href="MagneticField.1.html">man page</a> for usage
 * information.
 **********************************************************************/

#include <iostream>
#include <string>
#include <sstream>
#include <fstream>
#include <GeographicLib/MagneticModel.hpp>
#include <GeographicLib/MagneticCircle.hpp>
#include <GeographicLib/DMS.hpp>
#include <GeographicLib/Utility.hpp>

#if defined(_MSC_VER)
// Squelch warnings about constant conditional expressions
#  pragma warning (disable: 4127)
#endif

#include "MagneticField.usage"

int main(int argc, char* argv[]) {
  try {
    using namespace GeographicLib;
    typedef Math::real real;
    bool verbose = false;
    std::string dir;
    std::string model = MagneticModel::DefaultMagneticName();
    std::string istring, ifile, ofile, cdelim;
    char lsep = ';';
    real time = 0, lat = 0, h = 0;
    bool timeset = false, circle = false, rate = false;
    real hguard = 500000, tguard = 50;
    int prec = 1;

    for (int m = 1; m < argc; ++m) {
      std::string arg(argv[m]);
      if (arg == "-n") {
        if (++m == argc) return usage(1, true);
        model = argv[m];
      } else if (arg == "-d") {
        if (++m == argc) return usage(1, true);
        dir = argv[m];
      } else if (arg == "-t") {
        if (++m == argc) return usage(1, true);
        try {
          time = Utility::fractionalyear<real>(std::string(argv[m]));
          timeset = true;
          circle = false;
        }
        catch (const std::exception& e) {
          std::cerr << "Error decoding argument of " << arg << ": "
                    << e.what() << "\n";
          return 1;
        }
      } else if (arg == "-c") {
        if (m + 3 >= argc) return usage(1, true);
        try {
          time = Utility::fractionalyear<real>(std::string(argv[++m]));
          DMS::flag ind;
          lat = DMS::Decode(std::string(argv[++m]), ind);
          if (ind == DMS::LONGITUDE)
            throw GeographicErr("Bad hemisphere letter on latitude");
          if (!(std::abs(lat) <= 90))
            throw GeographicErr("Latitude not in [-90d, 90d]");
          h = Utility::num<real>(std::string(argv[++m]));
          timeset = false;
          circle = true;
        }
        catch (const std::exception& e) {
          std::cerr << "Error decoding argument of " << arg << ": "
                    << e.what() << "\n";
          return 1;
        }
      } else if (arg == "-r")
        rate = !rate;
      else if (arg == "-p") {
        if (++m == argc) return usage(1, true);
        try {
          prec = Utility::num<int>(std::string(argv[m]));
        }
        catch (const std::exception&) {
          std::cerr << "Precision " << argv[m] << " is not a number\n";
          return 1;
        }
      } else if (arg == "-T") {
        if (++m == argc) return usage(1, true);
        try {
          tguard = Utility::num<real>(std::string(argv[m]));
        }
        catch (const std::exception& e) {
          std::cerr << "Error decoding argument of " << arg << ": "
                    << e.what() << "\n";
          return 1;
        }
      } else if (arg == "-H") {
        if (++m == argc) return usage(1, true);
        try {
          hguard = Utility::num<real>(std::string(argv[m]));
        }
        catch (const std::exception& e) {
          std::cerr << "Error decoding argument of " << arg << ": "
                    << e.what() << "\n";
          return 1;
        }
      } else if (arg == "-v")
        verbose = true;
      else if (arg == "--input-string") {
        if (++m == argc) return usage(1, true);
        istring = argv[m];
      } else if (arg == "--input-file") {
        if (++m == argc) return usage(1, true);
        ifile = argv[m];
      } else if (arg == "--output-file") {
        if (++m == argc) return usage(1, true);
        ofile = argv[m];
      } else if (arg == "--line-separator") {
        if (++m == argc) return usage(1, true);
        if (std::string(argv[m]).size() != 1) {
          std::cerr << "Line separator must be a single character\n";
          return 1;
        }
        lsep = argv[m][0];
      } else if (arg == "--comment-delimiter") {
        if (++m == argc) return usage(1, true);
        cdelim = argv[m];
      } else if (arg == "--version") {
        std::cout
          << argv[0] << ": GeographicLib version "
          << GEOGRAPHICLIB_VERSION_STRING << "\n";
        return 0;
      } else {
        int retval = usage(!(arg == "-h" || arg == "--help"), arg != "--help");
        if (arg == "-h")
          std::cout<< "\nDefault magnetic path = \""
                   << MagneticModel::DefaultMagneticPath()
                   << "\"\nDefault magnetic name = \""
                   << MagneticModel::DefaultMagneticName()
                   << "\"\n";
        return retval;
      }
    }

    if (!ifile.empty() && !istring.empty()) {
      std::cerr << "Cannot specify --input-string and --input-file together\n";
      return 1;
    }
    if (ifile == "-") ifile.clear();
    std::ifstream infile;
    std::istringstream instring;
    if (!ifile.empty()) {
      infile.open(ifile.c_str());
      if (!infile.is_open()) {
        std::cerr << "Cannot open " << ifile << " for reading\n";
        return 1;
      }
    } else if (!istring.empty()) {
      std::string::size_type m = 0;
      while (true) {
        m = istring.find(lsep, m);
        if (m == std::string::npos)
          break;
        istring[m] = '\n';
      }
      instring.str(istring);
    }
    std::istream* input = !ifile.empty() ? &infile :
      (!istring.empty() ? &instring : &std::cin);

    std::ofstream outfile;
    if (ofile == "-") ofile.clear();
    if (!ofile.empty()) {
      outfile.open(ofile.c_str());
      if (!outfile.is_open()) {
        std::cerr << "Cannot open " << ofile << " for writing\n";
        return 1;
      }
    }
    std::ostream* output = !ofile.empty() ? &outfile : &std::cout;

    tguard = std::max(real(0), tguard);
    hguard = std::max(real(0), hguard);
    prec = std::min(10, std::max(0, prec));
    int retval = 0;
    try {
      const MagneticModel m(model, dir);
      if ((timeset || circle)
          && (!Math::isfinite<real>(time) ||
              time < m.MinTime() - tguard ||
              time > m.MaxTime() + tguard))
        throw GeographicErr("Time " + Utility::str(time) +
                            " too far outside allowed range [" +
                            Utility::str(m.MinTime()) + "," +
                            Utility::str(m.MaxTime()) + "]");
      if (circle
          && (!Math::isfinite<real>(h) ||
              h < m.MinHeight() - hguard ||
              h > m.MaxHeight() + hguard))
        throw GeographicErr("Height " + Utility::str(h/1000) +
                            "km too far outside allowed range [" +
                            Utility::str(m.MinHeight()/1000) + "km," +
                            Utility::str(m.MaxHeight()/1000) + "km]");
      if (verbose) {
        std::cerr << "Magnetic file: " << m.MagneticFile()      << "\n"
                  << "Name: "          << m.MagneticModelName() << "\n"
                  << "Description: "   << m.Description()       << "\n"
                  << "Date & Time: "   << m.DateTime()          << "\n"
                  << "Time range: ["
                  << m.MinTime() << ","
                  << m.MaxTime() << "]\n"
                  << "Height range: ["
                  << m.MinHeight()/1000 << "km,"
                  << m.MaxHeight()/1000 << "km]\n";
      }
      if ((timeset || circle) && (time < m.MinTime() || time > m.MaxTime()))
        std::cerr << "WARNING: Time " << time
                  << " outside allowed range ["
                  << m.MinTime() << "," << m.MaxTime() << "]\n";
      if (circle && (h < m.MinHeight() || h > m.MaxHeight()))
        std::cerr << "WARNING: Height " << h/1000
                  << "km outside allowed range ["
                  << m.MinHeight()/1000 << "km,"
                  << m.MaxHeight()/1000 << "km]\n";
      const MagneticCircle c(circle ? m.Circle(time, lat, h) :
                             MagneticCircle());
      std::string s, stra, strb;
      while (std::getline(*input, s)) {
        try {
          std::string eol("\n");
          if (!cdelim.empty()) {
            std::string::size_type m = s.find(cdelim);
            if (m != std::string::npos) {
              eol = " " + s.substr(m) + "\n";
              s = s.substr(0, m);
            }
          }
          std::istringstream str(s);
          if (!(timeset || circle)) {
            if (!(str >> stra))
              throw GeographicErr("Incomplete input: " + s);
            time = Utility::fractionalyear<real>(stra);
            if (time < m.MinTime() - tguard || time > m.MaxTime() + tguard)
              throw GeographicErr("Time " + Utility::str(time) +
                                  " too far outside allowed range [" +
                                  Utility::str(m.MinTime()) + "," +
                                  Utility::str(m.MaxTime()) +
                                  "]");
            if (time < m.MinTime() || time > m.MaxTime())
              std::cerr << "WARNING: Time " << time
                        << " outside allowed range ["
                        << m.MinTime() << "," << m.MaxTime() << "]\n";
          }
          real lon;
          if (circle) {
            if (!(str >> strb))
              throw GeographicErr("Incomplete input: " + s);
            DMS::flag ind;
            lon = DMS::Decode(strb, ind);
            if (ind == DMS::LATITUDE)
              throw GeographicErr("Bad hemisphere letter on " + strb);
            if (lon < -540 || lon >= 540)
              throw GeographicErr("Longitude " + strb +
                                  " not in [-540d, 540d)");
          } else {
            if (!(str >> stra >> strb))
              throw GeographicErr("Incomplete input: " + s);
            DMS::DecodeLatLon(stra, strb, lat, lon);
            h = 0;              // h is optional
            if (str >> h) {
              if (h < m.MinHeight() - hguard || h > m.MaxHeight() + hguard)
                throw GeographicErr("Height " + Utility::str(h/1000) +
                                    "km too far outside allowed range [" +
                                    Utility::str(m.MinHeight()/1000) + "km," +
                                    Utility::str(m.MaxHeight()/1000) + "km]");
              if (h < m.MinHeight() || h > m.MaxHeight())
                std::cerr << "WARNING: Height " << h/1000
                          << "km outside allowed range ["
                          << m.MinHeight()/1000 << "km,"
                          << m.MaxHeight()/1000 << "km]\n";
            }
            else
              str.clear();
          }
          if (str >> stra)
            throw GeographicErr("Extra junk in input: " + s);
          real bx, by, bz, bxt, byt, bzt;
          if (circle)
            c(lon, bx, by, bz, bxt, byt, bzt);
          else
            m(time, lat, lon, h, bx, by, bz, bxt, byt, bzt);
          real H, F, D, I, Ht, Ft, Dt, It;
          MagneticModel::FieldComponents(bx, by, bz, bxt, byt, bzt,
                                         H, F, D, I, Ht, Ft, Dt, It);

          *output << DMS::Encode(D, prec + 1, DMS::NUMBER) << " "
                  << DMS::Encode(I, prec + 1, DMS::NUMBER) << " "
                  << Utility::str<real>(H, prec) << " "
                  << Utility::str<real>(by, prec) << " "
                  << Utility::str<real>(bx, prec) << " "
                  << Utility::str<real>(-bz, prec) << " "
                  << Utility::str<real>(F, prec) << eol;
          if (rate)
            *output << DMS::Encode(Dt, prec + 1, DMS::NUMBER) << " "
                    << DMS::Encode(It, prec + 1, DMS::NUMBER) << " "
                    << Utility::str<real>(Ht, prec) << " "
                    << Utility::str<real>(byt, prec) << " "
                    << Utility::str<real>(bxt, prec) << " "
                    << Utility::str<real>(-bzt, prec) << " "
                    << Utility::str<real>(Ft, prec) << eol;
        }
        catch (const std::exception& e) {
          *output << "ERROR: " << e.what() << "\n";
          retval = 1;
        }
      }
    }
    catch (const std::exception& e) {
      std::cerr << "Error reading " << model << ": " << e.what() << "\n";
      retval = 1;
    }
    return retval;
  }
  catch (const std::exception& e) {
    std::cerr << "Caught exception: " << e.what() << "\n";
    return 1;
  }
  catch (...) {
    std::cerr << "Caught unknown exception\n";
    return 1;
  }
}
